src/com/xilinx/rapidwright/router/VersalClockRouting.java

```diff
@@ -0,0 +1,494 @@
+        Queue<NodeWithPrevAndCost> q = new PriorityQueue<>();
+        Set<Node> visited = new HashSet<>();
+        Set<PIP> allPIPs = new HashSet<>();
+        Set<NodeWithPrevAndCost> startingPoints = new HashSet<>();
+        startingPoints.add(new NodeWithPrevAndCost(vroute));
+        // Pattern: NODE_GLOBAL_VROUTE -> ... -> NODE_GLOBAL_VDISTR_LVL2 -> ... -> NODE_GLOBAL_VDISTR_LVL1 -> ... -> NODE_GLOBAL_VDISTR
+        Set<IntentCode> allowedIntentCodes = EnumSet.of(
+            IntentCode.NODE_GLOBAL_VDISTR,
+            IntentCode.NODE_GLOBAL_VDISTR_LVL1,
+            IntentCode.NODE_GLOBAL_VDISTR_LVL2,
+            IntentCode.NODE_GLOBAL_GCLK
+        );
+        nextClockRegion: for (ClockRegion cr : clockRegions) {
+            q.clear();
+            visited.clear();
+            q.addAll(startingPoints);
+            Tile crApproxCenterTile = cr.getApproximateCenter();
+            while (!q.isEmpty()) {
+                NodeWithPrevAndCost curr = q.poll();
+                IntentCode c = curr.getIntentCode();
+                ClockRegion currCR = curr.getTile().getClockRegion();
+                if (currCR != null && cr.getRow() == currCR.getRow() && c == IntentCode.NODE_GLOBAL_VDISTR) {
+                    // Only consider base wires
+                    if (getNodeStatus.apply(curr) == NodeStatus.INUSE) {
+                        startingPoints.add(curr);
+                    } else {
+                        List<Node> path = curr.getPrevPath();
+                        for (Node node : path) {
+                            startingPoints.add(new NodeWithPrevAndCost(node));
+                        }
+                        allPIPs.addAll(RouterHelper.getPIPsFromNodes(path));
+                    }
+                    crToVdist.put(cr, curr);
+                    continue nextClockRegion;
+                }
+
+                for (Node downhill : curr.getAllDownhillNodes()) {
+                    if (!allowedIntentCodes.contains(downhill.getIntentCode())) {
+                        continue;
+                    }
+                    if (!visited.add(downhill)) {
+                        continue;
+                    }
+                    int cost = downhill.getTile().getManhattanDistance(crApproxCenterTile);
+                    q.add(new NodeWithPrevAndCost(downhill, curr, cost));
+                }
+            }
+            throw new RuntimeException("ERROR: Couldn't route to distribution line in clock region " + cr);
+        }
+        clk.getPIPs().addAll(allPIPs);
+        return crToVdist;
+    }
+
+    /**
+     * Routes from a vertical distribution centroid to destination horizontal distribution lines
+     * in the clock regions provided.
+     * @param clk The current clock net
+     * @param crMap A map of target clock regions and their respective vertical distribution lines
+     * @return The List of nodes from the centroid to the horizontal distribution line.
```

Comment:
@WenhaoLin-AMD Can you update this javadoc line please?

Reply:
Have updated this part in https://github.com/Xilinx/RapidWright/pull/1102/commits/823f291b38eb59d5c650690939746d436cc1f290.